### PR TITLE
Strip impl if not re-exported and is doc(hidden)

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #112852.

r? @aDotInTheVoid

### CODE REVIEW

File: tests/rustdoc-json/reexport/issue-112852-dangling-trait-impl-id.rs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#![feature(no_core)]
```
Comment (aDotInTheVoid) [inline tests/rustdoc-json/reexport/issue-112852-dangling-trait-impl-id.rs:1]: NIT: this should be in the `impls` or `traits` folder, as there are no reexports

File: tests/rustdoc-json/reexport/issue-112852-dangling-trait-impl-id.rs [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+#![no_core]
+
+// @has "$.index[*][?(@.name=='PubTrait')]"
+pub trait PubTrait {}
+
```
Comment (aDotInTheVoid) [inline tests/rustdoc-json/reexport/issue-112852-dangling-trait-impl-id.rs:6]: It's worth being more explicit about what we're testing


```suggestion
// @!has "$.index[*][?(@.name == 'HiddenPubStruct')]"
// @!has "$.index[*][?(@.inner.impl)]"
```

**Comment by aDotInTheVoid** [comment]
What else is needed to close #112852? Just tests?

**Comment by GuillaumeGomez** [comment]
Yes, more tests.

File: tests/rustdoc-json/reexport/issue-112852-dangling-trait-impl-id.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply tests/rustdoc-json/reexport/issue-112852-dangling-trait-impl-id.rs:1]: I don't know why I put it there... Moving the location.

**Comment by GuillaumeGomez** [comment]
Applied comments.

**Comment by aDotInTheVoid** [comment]
> Applied comments.

IDK why GitHub highlighted it differently, but I also suggested adding `// @!has "$.index[*][?(@.inner.impl)]"`

> Yes, more tests.

Could they be added now? This code contains logic (that AFAIKT is right) for handling reexports, but if it's wrong, I'd like to know now, and not when doing a follow-up PR to add tests.

**Comment by GuillaumeGomez** [comment]
Sure I don't mind. I just have no idea what tests to add. If you have ideas, please don't hesitate. :)

**Comment by aDotInTheVoid** [comment]
> Sure I don't mind. I just have no idea what tests to add. If you have ideas, please don't hesitate. :)

[See here](https://github.com/rust-lang/rust/issues/112852#issue-1765882871)

> When this is fixed, we should add tests for these 2 cases, them with a pub use of the struct, them with --document-hidden-items, and them with --document-hidden-items and a pub use.

**Comment by GuillaumeGomez** [comment]
I added two more tests. To fix the `-3` one, I had to actually fix the `--document-hidden-items` option in rustdoc which was handled very poorly.

**Comment by GuillaumeGomez** [comment]
Fixed the CI errors.

**Comment by GuillaumeGomez** [comment]
Considering it got completely away from JSON (because of `--document-hidden-items`), I'll set @notriddle as reviewer.

r? @notriddle

**Review by notriddle — APPROVED** [review]

**Review by notriddle — CHANGES_REQUESTED** [review]

File: tests/rustdoc-json/impls/issue-112852-dangling-trait-impl-id.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#![feature(no_core)]
+#![no_core]
+
+// @!has "$.index[*][?(@.name == 'HiddenPubStruct')]"
+// @!has "$.index[*][?(@.inner.impl)]"
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply tests/rustdoc-json/impls/issue-112852-dangling-trait-impl-id.rs:5]: That's still not ideal (I really don't like `@!has`) but in JSON what you suggested is already much better. Going to update.

**Comment by GuillaumeGomez** [comment]
Updated!

**Pull request merged**